apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/observer/TrimObserver.java

```diff
@@ -36,12 +47,66 @@
     private final Object lock = new Object();
     private final Queue<Integer> trimHeights = new PriorityQueue<>(); // will sort heights from lowest to highest automatically
     private ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+    private BlockchainConfig blockchainConfig;
+    private PropertiesHolder propertiesHolder;
+    private Blockchain blockchain;
+    private Random random;
+    private boolean isShardingOff;
+    private final int maxRollback;
+
+    @Inject
+    public TrimObserver(TrimService trimService, BlockchainConfig blockchainConfig,
+                        PropertiesHolder propertiesHolder, Random random,
+                        Blockchain blockchain) {
+        this.trimService = Objects.requireNonNull(trimService, "trimService is NULL");
+        this.blockchainConfig = Objects.requireNonNull(blockchainConfig, "blockchainConfig is NULL");
+        this.propertiesHolder = Objects.requireNonNull(propertiesHolder, "propertiesHolder is NULL");
+        this.blockchain = Objects.requireNonNull(blockchain, "blockchain is NULL");
+        this.trimFrequency = Constants.DEFAULT_TRIM_FREQUENCY;
+        this.isShardingOff = this.propertiesHolder.getBooleanProperty("apl.noshardcreate", false);
+        if (random == null) {
+            this.random = new Random();
+        } else {
+            this.random = random;
+        }
+        this.maxRollback = this.propertiesHolder.getIntProperty("apl.maxRollback", 720);
+    }
 
     @PostConstruct
     void init() {
-        executorService.scheduleWithFixedDelay(this::processTrimEvent, 0, 2, TimeUnit.SECONDS);
+        if (this.blockchainConfig.getCurrentConfig().isShardingEnabled()
+                && this.blockchainConfig.getCurrentConfig().getShardingFrequency() > 0 && this.trimFrequency > 0
+                && this.blockchainConfig.getCurrentConfig().getShardingFrequency() < this.trimFrequency) {
+            String error = String.format(
+                    "SHARDING FREQUENCY ERROR: configured 'shard frequency value'=%d is LOWER then 'DEFAULT_TRIM_FREQUENCY'=%d",
+                    this.blockchainConfig.getCurrentConfig().getShardingFrequency(), this.trimFrequency);
+            log.error(error);
+            throw new RuntimeException(error);
+        }
+        // schedule first run with random delay in 2 sec range
+        executorService.schedule(taskToCall, ThreadLocalRandom.current().nextLong(
+                LONG_TIME_TWO_SECONDS - 1L) + 1L, TimeUnit.MILLISECONDS);
     }
 
+    /**
+     * Callable task for method to run. Next run is scheduled as soon as previous has finished
+     */
+    Callable<Void> taskToCall = new Callable<>() {
```

Comment:
Better to randomize by block height

Reply:
That will be useless rolling back.
Because around 'shard height point' many nodes will emit the same height, so a little gap by time is helpful here.